### PR TITLE
Add support for modifiers and allow to auto-activate `WheelZoomTool`

### LINKED ISSUE #10439: Activate wheel zoom by default with BokehJS
With BokehJS, I would like to have the wheel zoom for scrolling activated by default.
It is not the case by default. 

An option like: 
`active_scroll = "wheel_zoom" `
is not implemented.

But changing the line 30073 of bokeh.js file from
`return et == 'tap' || et == 'pan';`
to
`return et == 'tap' || et == 'pan' || et == 'scroll';`
activates the wheel zoom.

Am I breaking something doing this ? if not could it be added in next release.

Simple code to test this is availaible from https://jsfiddle.net/PBrockmann/ct6okdya/

### PR DESCRIPTION
**Pull request by mattpap**
Originally I started working on scroll support in `BoxAnnotation` per [issue #13646](https://github.com/bokeh/bokeh/issues/13646#issuecomment-2040446206), but I realized that I need to resolve the more generic issue of conditional handling of scroll events first, so that bokehjs doesn't interfere by with page scrolling, while allowing some way of easy access to scrolling functionality (like wheel zoom or wheel pan tools provide).

Thus this PR:

- [x] redesigns how tool event handlers are invoked and allows to indicated from such handlers if an event was handled or not (this has the added benefit of not using expensive signaling infrastructure to dispatch UI events to tool handlers)
- [x] adds support for `modifiers` to `WheelZoomTool` and `WheelPanTool`
- [x] adds support to `Toolbar` for auto-activation of such tools when `modifiers` are set
- [x] adds support for a visual indicator that a modifier is required to perform tool's function (e.g. an alert or a plot overlay)

This example shows `WheelZoomTool` configured with `modifiers = dict(ctrl=True)`, first being scrolled without the modifier and then with:

[Screencast from 09.04.2024 02:22:41.webm](https://github.com/bokeh/bokeh/assets/27475/1771e505-354d-49b4-b3ea-2f431bb36270)

fixes #10439

### CODE REVIEW

**Comment by jbednar** [comment]
Cool! Can you describe the details of how it "redesigns how tool event handlers are invoked"?

**Comment by mattpap** [comment]
> Cool! Can you describe the details of how it "redesigns how tool event handlers are invoked"?

The previous approach using signals didn't allow for returning a value from handlers (like `WheelZoomTool._scroll`), so this PR changes that, replacing signals with direct function calls.

File: bokehjs/src/lib/models/plots/plot_canvas.ts [hunk 1]
```diff
@@ -1190,4 +1217,20 @@ export class PlotView extends LayoutDOMView implements Renderable {
...
+    }
+    const el = new Div({children: [message]})
+    const timer = setTimeout(() => {
+      this._messages.delete(message)
+      this._notifications.elements = this._notifications.elements.filter((item) => item != el)
+    }, 2000)
+    this._messages.set(message, timer)
+    this._notifications.elements = [...this._notifications.elements, el]
+    logger.info(message)
+  }
```
Comment (bryevdv) [inline bokehjs/src/lib/models/plots/plot_canvas.ts:1235]: I expect many users would find something like this extremely useful if we can find a good way to expose it

File: src/bokeh/models/tools.py [hunk 2]
```diff
@@ -531,6 +534,20 @@ def __init__(self, *args, **kwargs) -> None:
     wheel pan tool will pan the plot along the x-axis.
     """)
 
+    modifiers = Modifiers(default={}, help="""
+    Allows to configure a combination of modifier keys, which need to
+    be pressed during the selected gesture for this tool to trigger.
```
Comment (bryevdv) [inline src/bokeh/models/tools.py:555]: I think an example code block for setting this property would be extremely helpful (here and elsewhere)

**Review by bryevdv — APPROVED** [review]

File: bokehjs/src/lib/models/plots/plot_canvas.ts (hunk 1 above)
Comment (mattpap) [reply bokehjs/src/lib/models/plots/plot_canvas.ts:1235]: I intend to generalize this in future PRs for 3.5. For now this implementation is just enough to make this PR viable.

Comment (mattpap) [reply bokehjs/src/lib/models/plots/plot_canvas.ts:1235]: I will probably use issue #6884 to follow up on this.

**Pull request merged**